### PR TITLE
Feature/PI-387 use fetch instead of request promise

### PR DESCRIPTION
**Pull request by Dynki**
In order to allow the portal to migrate to Vite, we had to patch code in the @rhino.fi/client-js package. This is not ideal and so this PR is to change client-js to allow the portal patches to be removed.

The following files have been updated.

1. @rhino.fi/client-js/src/api/deposit.js
2. @rhino.fi/client-js/src/api/estimatedNextBatchTime.js
3. @rhino.fi/client-js/src/api/eth/getGasStationPrice.js
4. @rhino.fi/client-js/src/api/fastWithdrawal.js
5. @rhino.fi/client-js/src/api/fastWithdrawalFee.js
6. @rhino.fi/client-js/src/api/getConfig.js
7. @rhino.fi/client-js/src/api/getGasPrice.js
8. @rhino.fi/client-js/src/api/getTickers.js
9. @rhino.fi/client-js/src/api/getTokenHolders.js
10. @rhino.fi/client-js/src/api/getTokenLiquidityLeft.js
11. @rhino.fi/client-js/src/api/getTokenSaleStartEnd.js
12. @rhino.fi/client-js/src/api/ledger/deposit.js
13. @rhino.fi/client-js/src/api/ledger/transferUsingVaultIdAndStarkKey.js
14. @rhino.fi/client-js/src/api/ledger/withdraw.js
15. @rhino.fi/client-js/src/api/submitMarketOrder.js
16. @rhino.fi/client-js/src/api/submitOrder.js
17. @rhino.fi/client-js/src/api/transferUsingVaultIdAndStarkKey.js
18. @rhino.fi/client-js/src/api/withdraw.js
19. @rhino.fi/client-js/src/lib/dvf/post-generic.js

- The post-generic file has been updated to use `request` from `dvf_utils` 
- The remaining 18 files have been updated to use either the `post-generic` or `get-generic` functions

### CODE REVIEW

File: src/api/estimatedNextBatchTime.js [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
+const get = require('../lib/dvf/get-generic')
 
 module.exports = async dvf => {
   // avoid browser cache with timestamp as querystring
   const t = Date.now()
 
   const url = `${dvf.config.api}/v1/trading/r/estimatedNextBatchTime?t=${t}`
   try {
-    const data = await get(url)
+    const data = await get(dvf, url)
```
Comment (adrian-gierakowski) [inline src/api/estimatedNextBatchTime.js:9]: generic-get prepends `dvf.config.api` to the 2nd arg so if you want to use it we should remove do:

```js
const url = `/v1/trading/r/estimatedNextBatchTime?t=${t}`
```

https://github.com/rhinofi/client-js/blob/a51ee785cf74e2de056cc7480d92b248c4b560a6/src/lib/dvf/get-generic.js#L6

File: src/api/estimatedNextBatchTime.js [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
     return JSON.parse(data)
   }
   catch(e) {
```
Comment (adrian-gierakowski) [inline src/api/estimatedNextBatchTime.js:12]: the catch here is unnecessary, so pulling all the comment together the function should become:

```js
module.exports = async dvf => {
  // avoid browser cache with timestamp as querystring
  const t = Date.now()
  const url = `/v1/trading/r/estimatedNextBatchTime?t=${t}`
  return get(dvf, url)
}
```

File: src/api/eth/getGasStationPrice.js [hunk 3]
```diff
@@ -1,12 +1,11 @@
...
-
+const get = require('../../lib/dvf/get-generic')
 /**
  * Provides a safe average gas price
  */
 
 module.exports = async (dvf)  => {
   try {
-    const res = await get(`${dvf.config.gasApi}/json/ethgasAPI.json?api-key=${dvf.config.gasStationApiKey || ''}`)
+    const res = await get(dvf, `${dvf.config.gasApi}/json/ethgasAPI.json?api-key=${dvf.config.gasStationApiKey || ''}`)
```
Comment (adrian-gierakowski) [inline src/api/eth/getGasStationPrice.js:8]: given the previous comments, we cannot use get-generic here as it prepend `dvf.config.api`  and we need `dvf.config.gasApi` here, so let's switch to plain `request.get`  from `@rhino.fi/dvf-utils`

File: src/api/fastWithdrawal.js [hunk 4]
```diff
@@ -1,7 +1,7 @@
-const { post } = require('request-promise')
+const post = require('../lib/dvf/post-generic')
 
 module.exports = async (dvf, withdrawalData) => {
   const url = dvf.config.api + '/v1/trading/w/fastWithdrawal'
```
Comment (adrian-gierakowski) [inline src/api/fastWithdrawal.js:4]: as above 
```suggestion
  const url = '/v1/trading/w/fastWithdrawal'
```

File: src/api/fastWithdrawalFee.js [hunk 5]
```diff
@@ -1,7 +1,7 @@
-const { get } = require('request-promise')
+const get = require('../lib/dvf/get-generic')
 
 // TODO: Deprecated (CHAIN-719)
 module.exports = async (dvf, token) => {
   const url = dvf.config.api + '/v1/trading/r/fastWithdrawalFee'
```
Comment (adrian-gierakowski) [inline src/api/fastWithdrawalFee.js:5]: same again

File: src/api/getTickers.js [hunk 6]
```diff
@@ -1,10 +1,10 @@
...
+const get = require('../lib/dvf/get-generic')
 
 module.exports = async (dvf, symbols) => {
   if (symbols.constructor !== Array) {
     symbols = [symbols]
   }
 
-  const response = await get(`${dvf.config.api}/market-data/tickers?symbols=${symbols.join(',')}`)
+  const response = await get(dvf, `${dvf.config.api}/market-data/tickers?symbols=${symbols.join(',')}`)
   return JSON.parse(response)
```
Comment (adrian-gierakowski) [inline src/api/getTickers.js:9]:
```suggestion
  return get(dvf, `/market-data/tickers?symbols=${symbols.join(',')}`)
```

File: src/api/getConfig.js [hunk 7]
```diff
@@ -1,9 +1,9 @@
-const { post } = require('request-promise')
+const post = require('../lib/dvf/post-generic')
 
 module.exports = async dvf => {
   const url = dvf.config.api + '/v1/trading/r/getConf'
```
Comment (adrian-gierakowski) [inline src/api/getConfig.js:4]: as above

File: src/api/getTokenHolders.js [hunk 8]
```diff
@@ -1,8 +1,9 @@
-const { get } = require('request-promise')
+const get = require('../lib/dvf/get-generic')
 
 module.exports = async (dvf, token) => {
   if (token) {
     const response = await get(
+      dvf,
       `${dvf.config.api}/v1/trading/r/getTokenHolders?token=${token}`
```
Comment (adrian-gierakowski) [inline src/api/getTokenHolders.js:7]: as above

File: src/api/getTokenLiquidityLeft.js [hunk 9]
```diff
@@ -1,8 +1,9 @@
-const { get } = require('request-promise')
+const get = require('../lib/dvf/get-generic')
 
 module.exports = async (dvf, token) => {
   if (token) {
     const response = await get(
+      dvf,
       `${dvf.config.api}/v1/trading/r/getTokenLiquidityLeft?token=${token}`
```
Comment (adrian-gierakowski) [inline src/api/getTokenLiquidityLeft.js:7]: as above

File: src/api/submitOrder.js [hunk 10]
```diff
@@ -67,5 +67,5 @@ module.exports = async (dvf, orderData) => {
       : dvf.createOrderPayload(value)
   )
 
-  return post(dvf.config.api + '/v1/trading/w/submitOrder', { json })
+  return post(dvf, dvf.config.api + '/v1/trading/w/submitOrder', { json })
```
Comment (adrian-gierakowski) [inline src/api/submitOrder.js:70]: as above

File: src/api/transferUsingVaultIdAndStarkKey.js [hunk 11]
```diff
@@ -1,8 +1,8 @@
-const { post } = require('request-promise')
+const post = require('../lib/dvf/post-generic')
 
 module.exports = async (dvf, transferData, feeRecipient) => {
   const url = dvf.config.api + '/v1/trading/w/transfer'
```
Comment (adrian-gierakowski) [inline src/api/transferUsingVaultIdAndStarkKey.js:4]: as above

File: src/api/withdraw.js [hunk 12]
```diff
@@ -55,5 +55,5 @@ module.exports = async (dvf, token, amount, starkPrivateKey) => {
     expireTime
   }
   //console.log({ data })
-  return post(url, { json: data })
+  return post(dvf, url, { json: data })
```
Comment (adrian-gierakowski) [inline src/api/withdraw.js:58]: as above

**Review by adrian-gierakowski — CHANGES_REQUESTED** [review]

**Review by adrian-gierakowski — COMMENTED** [review]
did some more testing and it turns out that dvf-utils request has slightly different interface then request-promise so `src/lib/dvf/post-generic.js` shoud be changed to:

```js
const { request } = require('@rhino.fi/dvf-utils')
const _omitBy = require('lodash/omitBy')
const _isNil = require('lodash/isNil')

module.exports = async (dvf, endpoint, json = {}, headers = {}) => {
  const url = dvf.config.api + endpoint

  const options = {
    headers,
    // removes null and undefined values
    data: _omitBy(json, _isNil)
  }

  return request.post(url, options)
}
```

File: src/api/eth/getGasStationPrice.js (hunk 3 above)
Comment (Dynki) [reply src/api/eth/getGasStationPrice.js:8]: Do we need to parse the response in the following line then? As it looks like we don't, but I'm not sure

`dvf.config.defaultGasPrice = parseInt((JSON.parse(res).average * 1.25 *100000000))`

Comment (adrian-gierakowski) [reply src/api/eth/getGasStationPrice.js:8]: I guess not, If you remove the JSON.parse I can test to confirm

**Pull request merged**